### PR TITLE
[TACHYON-1415] Check - Add configuration properties for permission check usage

### PR DESCRIPTION
**Pull request by dongche**
JIRA: https://tachyon.atlassian.net/browse/TACHYON-1415

Add configuration properties and an almost empty class `FileSystemPermissionChecker` to use it. The check method impl will be in TACHYON-1416.

@apc999 , @sundapeng

### CODE REVIEW

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+  /** The owner of root directory. */
+  private static String sFileSystemOwner;
+
+  /** The super group of Tachyon file system. All users in this group has super permission. */
+  private static String sFileSystemSuperGroup;
+
+  /**
+   * Initialize the permission related property of the whole Tachyon file system.
+   * @param permissionCheckEnabled whether permission checking is enabled
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: and an empty line between the method-level comment and parameter descriptions

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  /**
+   * TODO(dong): add static methods to check access permission.
+   * The input is User and its Groups, requested Permission and INodes (traverse the Path).
+   *
+   * The initialized static attributes will be used in the checking logic to bypass checking.
+   * Then User, Group, and Action will be compared to those of INodes.
+   *
+   * It will return if check passed, and throw exception if check failed.
+   */
+  public static void checkPermission(User user, List<String> groups, FileSystemAction action,
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: add javadoc for the arguments

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    sFileSystemOwner = owner;
+    sFileSystemSuperGroup = superGroup;
+  }
+
+  /**
+   * TODO(dong): add static methods to check access permission.
+   * The input is User and its Groups, requested Permission and INodes (traverse the Path).
+   *
+   * The initialized static attributes will be used in the checking logic to bypass checking.
+   * Then User, Group, and Action will be compared to those of INodes.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: s/INodes/inodes

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  public static void checkPermission(User user, List<String> groups, FileSystemAction action,
+      List<Inode> iNodes) throws IOException {
+    if (!sPermissionCheckEnabled) {
+      return;
+    }
+
+    if (sFileSystemOwner.equals(user.getName()) || groups.contains(sFileSystemSuperGroup)) {
+      return;
+    }
+    // Impl in TACHYON-1416
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: Should this be: `TODO(dong): Implemented TACHYON-1416`?

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.List;
+
+import tachyon.master.file.meta.Inode;
+import tachyon.security.User;
+import tachyon.security.authorization.FileSystemAction;
+
+/**
+ * Base class to provide permission check logic.
+ */
+public class FileSystemPermissionChecker {
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: should this class be final?

File: servers/src/test/java/tachyon/master/file/meta/InodeTreeTest.java [hunk 6]
```diff
@@ -79,7 +83,13 @@ public void before() throws IOException {
     mTree = new InodeTree(blockMaster, directoryIdGenerator, mountTable);
 
     blockMaster.start(true);
+
+    TachyonConf conf = new TachyonConf();
+    conf.set(Constants.SECURITY_AUTHORIZATION_PERMISSION_ENABLED, "true");
+    conf.set(Constants.SECURITY_AUTHORIZATION_PERMISSION_SUPERGROUP, "superGroup");
```
Comment (jsimsa) [inline servers/src/test/java/tachyon/master/file/meta/InodeTreeTest.java]: replacing the default supergroup with superGroup is subtle ... let's use test-supergroup instead

**Comment by jsimsa** [comment]
@dongche thanks for the PR, left some comments

**Comment by dongche** [comment]
@jsimsa Thanks for your review. Patch updated to address comments.

File: common/src/main/java/tachyon/Constants.java [hunk 7]
```diff
@@ -364,5 +364,10 @@
   public static final short DEFAULT_TFS_FULL_PERMISSION = (short) 0777;
   public static final short FILE_DIR_PERMISSION_DIFF = (short) 0111;
 
+  public static final String SECURITY_AUTHORIZATION_PERMISSION_ENABLED =
+      "tachyon.security.authorization.permission.enabled";
```
Comment (jsimsa) [inline common/src/main/java/tachyon/Constants.java]: These properties should also be document in tachyon/docs/Configuration-Settings.md

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  private static boolean sPermissionCheckEnabled;
+
+  /** The owner of root directory. */
+  private static String sFileSystemOwner;
+
+  /** The super group of Tachyon file system. All users in this group has super permission. */
+  private static String sFileSystemSuperGroup;
+
+  /**
+   * Initialize the permission related property of the whole Tachyon file system.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: s/Initialize/Initializes/

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+   */
+  public static void initializeFileSystem(boolean permissionCheckEnabled, String owner,
+      String superGroup) {
+    sPermissionCheckEnabled = permissionCheckEnabled;
+    sFileSystemOwner = owner;
+    sFileSystemSuperGroup = superGroup;
+  }
+
+  /**
+   * This method check access permission.
```
Comment (jsimsa) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: s/check/checks

**Comment by jsimsa** [comment]
just two more minor comments ... other than that it LGTM

**Comment by jsimsa** [comment]
@dongche thanks

File: common/src/main/java/tachyon/Constants.java (hunk 7 above)
Comment (dongche) [inline common/src/main/java/tachyon/Constants.java]: Oh, yes. These security related properties are all missed there... Both the finished authentication properties, and these authorization properties in development.

After go through the `Configuration-Settings.md`, I am not sure which category to put them in (some are common, but some are master specific), or create a new category say `Security`. Also the property names might need changes for consistence.

So shall we open a new JIRA to track this?

**Comment by dongche** [comment]
@jsimsa Patch updated. Thanks for the comment about Configuration-Settings! I get some questions in the reply, which need to be considered. So could we do that changes in a follow-up JIRA?

**Comment by jsimsa** [comment]
@dongche addressing the configuration documentation in a separate PR sounds good, please create JIRA for this and add a TODO into this PR ... also creating a separate "Security" section for security related properties in the documentation makes sense ... thanks

LGTM

File: common/src/main/java/tachyon/Constants.java (hunk 7 above)
Comment (yupeng9) [inline common/src/main/java/tachyon/Constants.java]: Sounds good. Please do.

File: servers/src/main/java/tachyon/master/file/meta/InodeTree.java [hunk 10]
```diff
@@ -117,6 +118,15 @@ public void initializeRoot(PermissionStatus rootPermissionStatus) {
               .setPermissionStatus(rootPermissionStatus).setParentId(NO_PARENT).build();
       mInodes.add(mRoot);
       mCachedInode = mRoot;
+      FileSystemPermissionChecker.initializeFileSystem(
+          MasterContext.getConf().getBoolean(Constants.SECURITY_AUTHORIZATION_PERMISSION_ENABLED),
+          mRoot.getUserName(),
+          MasterContext.getConf().get(Constants.SECURITY_AUTHORIZATION_PERMISSION_SUPERGROUP));
+    } else {
+      FileSystemPermissionChecker.initializeFileSystem(
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/file/meta/InodeTree.java]: This call looks identical to the one in other branch. Can you pull it out then?

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Base class to provide permission check logic.
+ */
+public final class FileSystemPermissionChecker {
+  private static boolean sPermissionCheckEnabled;
+
+  /** The owner of root directory. */
+  private static String sFileSystemOwner;
+
+  /** The super group of Tachyon file system. All users in this group has super permission. */
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: has -> have

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  private static String sFileSystemSuperGroup;
+
+  /**
+   * Initializes the permission related property of the whole Tachyon file system.
+   *
+   * @param permissionCheckEnabled whether permission checking is enabled
+   * @param owner the user of root directory, who is seen as the super user
+   * @param superGroup the super group of the whole Tachyon file system
+   */
+  public static void initializeFileSystem(boolean permissionCheckEnabled, String owner,
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: Should this be synchronized?

File: servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  public static void checkPermission(User user, List<String> groups, FileSystemAction action,
+      List<Inode> inodes) throws IOException {
+    if (!sPermissionCheckEnabled) {
+      return;
+    }
+
+    if (sFileSystemOwner.equals(user.getName()) || groups.contains(sFileSystemSuperGroup)) {
+      return;
+    }
+    // TODO(dong): Implement TACHYON-1416
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: How difficult would it be to implement TACHYON-1416? Right now this method does not do any checking.

Comment (dongche) [inline servers/src/main/java/tachyon/master/permission/FileSystemPermissionChecker.java]: Not too much. Split them just for keeping the PR's size and reviewing easily. 
The checking logic and test are done locally and will be submitted after this PR.

File: common/src/main/java/tachyon/Constants.java (hunk 7 above)
Comment (dongche) [inline common/src/main/java/tachyon/Constants.java]: Create https://tachyon.atlassian.net/browse/TACHYON-1462 to track this.

**Comment by dongche** [comment]
@yupeng9 Thanks for your review. Patch updated to address comments.

**Comment by jsimsa** [comment]
LGTM

**Comment by apc999** [comment]
+1 LGTM

**Pull request merged**